pkg/gadgets/tcpconnect/gadget.go

```diff
@@ -0,0 +1,147 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tcpconnect
+
+import (
+	"encoding/json"
+	"fmt"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/tcpconnect/tracer"
+
+	coretracer "github.com/kinvolk/inspektor-gadget/pkg/gadgets/tcpconnect/tracer/core"
+	standardtracer "github.com/kinvolk/inspektor-gadget/pkg/gadgets/tcpconnect/tracer/standard"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/tcpconnect/types"
+
+	gadgetv1alpha1 "github.com/kinvolk/inspektor-gadget/pkg/apis/gadget/v1alpha1"
+)
+
+type Trace struct {
+	resolver gadgets.Resolver
+
+	started bool
+	tracer  tracer.Tracer
+}
+
+type TraceFactory struct {
+	gadgets.BaseFactory
+}
+
+func NewFactory() gadgets.TraceFactory {
+	return &TraceFactory{
+		BaseFactory: gadgets.BaseFactory{DeleteTrace: deleteTrace},
+	}
+}
+
+func (f *TraceFactory) Description() string {
+	return `tcpconnect traces connect() system calls`
+}
+
+func (f *TraceFactory) OutputModesSupported() map[string]struct{} {
+	return map[string]struct{}{
+		"Stream": {},
+	}
+}
+
+func deleteTrace(name string, t interface{}) {
+	trace := t.(*Trace)
+	if trace.tracer != nil {
+		trace.tracer.Stop()
+	}
+}
+
+func (f *TraceFactory) Operations() map[string]gadgets.TraceOperation {
+	n := func() interface{} {
+		return &Trace{
+			resolver: f.Resolver,
+		}
+	}
+
+	return map[string]gadgets.TraceOperation{
+		"start": {
+			Doc: "Start tcpconnect gadget",
+			Operation: func(name string, trace *gadgetv1alpha1.Trace) {
+				f.LookupOrCreate(name, n).(*Trace).Start(trace)
+			},
+		},
+		"stop": {
+			Doc: "Stop tcpconnect gadget",
+			Operation: func(name string, trace *gadgetv1alpha1.Trace) {
+				f.LookupOrCreate(name, n).(*Trace).Stop(trace)
+			},
+		},
+	}
+}
+
+func (t *Trace) Start(trace *gadgetv1alpha1.Trace) {
+	if t.started {
+		gadgets.CleanupTraceStatus(trace)
+		trace.Status.State = "Started"
+		return
+	}
+
+	traceName := gadgets.TraceName(trace.ObjectMeta.Namespace, trace.ObjectMeta.Name)
+
+	eventCallback := func(event types.Event) {
+		r, err := json.Marshal(event)
+		if err != nil {
+			log.Warnf("Gadget %s: error marshalling event: %s", trace.Spec.Gadget, err)
+			return
+		}
+		t.resolver.PublishEvent(traceName, string(r))
+	}
+
+	var err error
+
+	config := &tracer.Config{
+		MountnsMap: gadgets.TracePinPath(trace.ObjectMeta.Namespace, trace.ObjectMeta.Name),
+	}
+	t.tracer, err = coretracer.NewTracer(config, t.resolver, eventCallback, trace.Spec.Node)
+	if err != nil {
+		// TODO: The following line causes the client to fail.
```

Comment:
Is it still the case now since #527 was merged?

Reply:
You're right. I forgot to update this and the existing gadgets. Added a commit to handle it.